### PR TITLE
Size and density properties

### PR DESCRIPTION
**Pull request by nils-werner**
NumPy has a `.size` property that gives you the number of elements in an array. 

Additionally for sparse arrays a `.density` property may be handy, which is the ratio of nonzero elements to total elements in the array (a value between `0` and `1`).

This PR implements both properties for `COO` arrays

### CODE REVIEW

**Comment by hameerabbasi** [comment]
Seems fine, however it would be really nice to have docstrings for these, and if possible, tests.

**Comment by nils-werner** [comment]
Added them

**Comment by hameerabbasi** [comment]
Ignore the Travis CI failure. Numpy 1.14 changed the way it converts arrays to strings, and that's causing a doctests failure. I've addressed it over in #43.

**Comment by nils-werner** [comment]
Yup, I already noticed that

File: sparse/tests/test_core.py [hunk 1]
```diff
@@ -875,3 +875,13 @@ def test_scalar_shape_construction():
     s = COO(coords, x, shape=5)
 
     assert_eq(x, s)
+
+
+def test_density():
+    s = sparse.random((20, 30, 40), density=0.1)
+    assert 0.09 < s.density < 0.11
```
Comment (hameerabbasi) [inline sparse/tests/test_core.py]: It might be wise to use `np.isclose` here.

File: sparse/core.py [hunk 2]
```diff
@@ -253,6 +253,54 @@ def nnz(self):
...
+        COO.size : Number of elements.
+        COO.nnz : Number of nonzero elements.
+
+        Examples
+        --------
+        >>> x = np.zeros((10, 10))
+        >>> x[0, :] = 1
+        >>> s = COO.from_numpy(x)
+        >>> s.density
+        0.10000000000000001
```
Comment (hameerabbasi) [inline sparse/core.py]: This might fluctuate in different system floating-point implementations. Perhaps `np.isclose(density, 0.1)`?

**Comment by mrocklin** [comment]
Or perhaps change 10 to 8 above so that the result is evenly divisible by
two

**Comment by nils-werner** [comment]
@mrocklin that's clever! Done.

**Comment by hameerabbasi** [comment]
This looks ready. If you rebase, I can do a final review and merge.

**Comment by nils-werner** [comment]
Did you see my [code comment](https://github.com/mrocklin/sparse/pull/69/files/a3adb4c816b6bfd71de8d8966f3cb87ac3642a73)?

**Comment by hameerabbasi** [comment]
Hmmm, that's strange, it isn't showing any of your comments in that link, or on any commits.

**Comment by nils-werner** [comment]
Aha! after commenting on lines you have to also open the code review thingy.

File: sparse/core.py [hunk 3]
```diff
@@ -1277,7 +1325,7 @@ def astype(self, dtype, out=None):
 
     def maybe_densify(self, allowed_nnz=1e3, allowed_fraction=0.25):
         """ Convert to a dense numpy array if not too costly.  Err othrewise """
-        if reduce(operator.mul, self.shape) <= allowed_nnz or self.nnz >= np.prod(self.shape) * allowed_fraction:
+        if self.size <= allowed_nnz or self.density >= allowed_fraction:
```
[the remark this answers is not in the document]
Comment (hameerabbasi) [reply sparse/core.py]: It actually does make sense. If the density is low, then the cost of a dense array is higher compared to that of a sparse array, and we should avoid densifying.

If, however, density is high, the overhead of a sparse matrix far outweighs the benefits and we should densify.

[the remark this answers is not in the document]
Comment (nils-werner) [reply sparse/core.py]: True. Maybe I got confused that one `allowed` means `max` and the other one means `min`... What about renaming them to

    maybe_densify(self, max_nnz=1e3, min_density=0.25)

[the remark this answers is not in the document]
Comment (hameerabbasi) [reply sparse/core.py]: I would make sure that no tests break either here or in Dask before doing that. We can change unit tests here... But it's best to send a PR to Dask if any tests break there.

**Comment by hameerabbasi** [comment]
Is `NotImplementedError` the best way to show that densfication is forbidden? What would be the right exception here?

File: sparse/core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hameerabbasi) [reply sparse/core.py]: Let's change it anyway and I'll send a PR to Dask if anything fails for them.

**Comment by nils-werner** [comment]
I would say `ValueError` is the right one, as it's the function arguments that make the operation fail. A `NotImplementedError` sounds more like "whatever you pass to this function, it will never succeed".

File: sparse/core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nils-werner) [reply sparse/core.py]: Actually, shouldn't `max_nnz` be actually named `max_size`? We are not comparing it to `.nnz` anyways...

**Comment by hameerabbasi** [comment]
Best to switch that to a `ValueError` then.

File: sparse/core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hameerabbasi) [reply sparse/core.py]: Yes, `max_size` is a good idea as well.

File: sparse/core.py [hunk 4]
```diff
@@ -2412,19 +2412,19 @@ def maybe_densify(self, allowed_nnz=1000, allowed_fraction=0.25):
...
         an error.
 
         >>> x = np.zeros((5, 5), dtype=np.uint8)
         >>> x[2, 2] = 1
         >>> s = COO.from_numpy(x)
-        >>> s.maybe_densify(allowed_nnz=5, allowed_fraction=0.25)
+        >>> s.maybe_densify(max_size=5, min_fraction=0.25)
         Traceback (most recent call last):
             ...
         NotImplementedError: Operation would require converting large sparse array to dense
```
Comment (hameerabbasi) [inline sparse/core.py]: You need to update this line in the docstring. [See here.](https://docs.python.org/3/library/doctest.html#what-about-exceptions)

File: sparse/core.py [hunk 5]
```diff
@@ -2381,17 +2381,17 @@ def astype(self, dtype, out=None):
         assert out is None
         return self.elemwise(np.ndarray.astype, dtype)
 
-    def maybe_densify(self, allowed_nnz=1000, allowed_fraction=0.25):
+    def maybe_densify(self, max_size=1000, min_fraction=0.25):
```
Comment (hameerabbasi) [inline sparse/core.py]: Best keep this as `min_density`, it's clearer that it represents the density.

File: sparse/core.py [hunk 6]
```diff
@@ -2364,25 +2412,23 @@ def maybe_densify(self, allowed_nnz=1000, allowed_fraction=0.25):
...
         an error.
 
         >>> x = np.zeros((5, 5), dtype=np.uint8)
         >>> x[2, 2] = 1
         >>> s = COO.from_numpy(x)
-        >>> s.maybe_densify(allowed_nnz=5, allowed_fraction=0.25)
+        >>> s.maybe_densify(max_size=5, min_density=0.25)
         Traceback (most recent call last):
             ...
         NotImplementedError: Operation would require converting large sparse array to dense
```
Comment (hameerabbasi) [inline sparse/core.py]: Line 2425...

```
NotImplementedError: Operation would require converting large sparse array to dense
```

Need to change this to `ValueError`.

Comment (hameerabbasi) [reply sparse/core.py]: It's causing doctests to fail. :)

Comment (nils-werner) [reply sparse/core.py]: Currently there are a million tests failing on `master`, it is hard to spot my own mistakes...

Comment (hameerabbasi) [reply sparse/core.py]: That's strange, [Travis shows everything is working on master](https://travis-ci.org/mrocklin/sparse/branches).

**Comment by mrocklin** [comment]
> Is NotImplementedError the best way to show that densfication is forbidden? What would be the right exception here?

To me `NotImplementedError` means "We have not *yet* implemented this, but it is a reasonable request".  If it is a case of a type not being supported like "we don't think that densifying scipy.sparse matrices is a good idea" then I would raise a TypeError.  If it is a case of the values being inappropriate like "sometimes we do densify sparse arrays, but this particular array is very very sparse" then I would use a ValueError.

File: sparse/core.py (hunk 6 above)
Comment (nils-werner) [reply sparse/core.py]: I checked, they only fail on Numpy 1.12. Latest Numpy works fine...

Comment (hameerabbasi) [reply sparse/core.py]: That is really weird, I tested with Numpy 1.13.3 on my local system and it still works fine.

**Comment by hameerabbasi** [comment]
So, we all agree that it should be `ValueError` in `maybe_densify` then. :-)

**Comment by hameerabbasi** [comment]
Are you still making changes here? If not, this gets a +1 from me.

**Comment by nils-werner** [comment]
No, I'm done.

**Comment by hameerabbasi** [comment]
A rebase would be nice.

**Pull request merged**